### PR TITLE
Save per-iteration Output Tensor Results

### PR DESCRIPTION
**Pull request by laks0209**
* Command line argument -saveTensorData dumps First or All iteration Output Tensor results to csv files
* Output tensor corresponding to each iteration is saved in separate csv
files
* Summary.csv file contains the final result of each iteration, hash of
output tensor and pointer to the dumped output tensor file

### CODE REVIEW

File: Tools/WinMLRunner/WinMLRunner.sln [hunk 1]
```diff
@@ -10,38 +10,62 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WinMLRunnerTest", "..\..\Te
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8} = {81EA9CC6-8A26-4583-B1A4-84740EF815C8}
 	EndProjectSection
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ClassLibrary1", "..\ClassLibrary1\ClassLibrary1.csproj", "{12E5A5A7-E32C-4D2E-84DC-E937BE0A9DA8}"
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.sln]: Was a Csharp project supposed to be added to this solution in the PR with the name: ClassLibrary1?

File: Tools/WinMLRunner/WinMLRunner.vcxproj [hunk 2]
```diff
@@ -153,7 +154,7 @@
     </ClCompile>
     <Link>
       <SubSystem>Console</SubSystem>
-      <AdditionalDependencies>dxgi.lib;d3d12.lib;windowsapp.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>dxgi.lib;d3d12.lib;windowsapp.lib;mscoree.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.vcxproj]: Why is mscoree.lib needed as a dependency? From my understanding this is related to .NET Framework

File: Tools/WinMLRunner/WinMLRunner.vcxproj [hunk 3]
```diff
@@ -70,6 +70,7 @@
     <UseDebugLibraries>true</UseDebugLibraries>
     <PlatformToolset>v141</PlatformToolset>
     <CharacterSet>Unicode</CharacterSet>
+    <CLRSupport>false</CLRSupport>
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.vcxproj]: Why do we need to specify CLRSupport? Isn't this related to .NET?

**Comment by laks0209** [comment]
Made modifications according to the per-iteration performance dump. Now, the summary.csv file contains  the per-iteration performance results and the final result. Please review the latest commit.

File: Tools/WinMLRunner/WinMLRunner.sln [hunk 4]
```diff
@@ -12,31 +12,37 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WinMLRunnerTest", "..\..\Te
...
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|ARM64.Build.0 = Release|ARM64
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|x64.ActiveCfg = Release|x64
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|x64.Build.0 = Release|x64
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|x86.ActiveCfg = Release|Win32
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|x86.Build.0 = Release|Win32
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|Any CPU.ActiveCfg = Debug|Win32
 		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|ARM64.ActiveCfg = Debug|Win32
-		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.ActiveCfg = Debug|x64
-		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.Build.0 = Debug|x64
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.ActiveCfg = Release|x64
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.sln]: When you change this, it builds release when debug is specified.

File: Tools/WinMLRunner/WinMLRunner.sln [hunk 5]
```diff
@@ -12,31 +12,37 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WinMLRunnerTest", "..\..\Te
...
 		Debug|ARM64 = Debug|ARM64
 		Debug|x64 = Debug|x64
 		Debug|x86 = Debug|x86
+		Release|Any CPU = Release|Any CPU
 		Release|ARM64 = Release|ARM64
 		Release|x64 = Release|x64
 		Release|x86 = Release|x86
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|Any CPU.ActiveCfg = Debug|Win32
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.sln]: When we specify Any CPU Debug, it'll build to Win32 Debug

File: Tools/WinMLRunner/WinMLRunner.sln [hunk 6]
```diff
@@ -12,31 +12,37 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WinMLRunnerTest", "..\..\Te
...
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|ARM64.ActiveCfg = Debug|ARM64
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|ARM64.Build.0 = Debug|ARM64
-		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|x64.ActiveCfg = Debug|x64
-		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|x64.Build.0 = Debug|x64
+		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Debug|x64.ActiveCfg = Release|x64
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.sln]: When you change this, it builds release when debug is specified.

File: Tools/WinMLRunner/WinMLRunner.sln [hunk 7]
```diff
@@ -12,31 +12,37 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WinMLRunnerTest", "..\..\Te
...
 		{81EA9CC6-8A26-4583-B1A4-84740EF815C8}.Release|x86.Build.0 = Release|Win32
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|Any CPU.ActiveCfg = Debug|Win32
 		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|ARM64.ActiveCfg = Debug|Win32
-		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.ActiveCfg = Debug|x64
-		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.Build.0 = Debug|x64
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.ActiveCfg = Release|x64
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x64.Build.0 = Release|x64
 		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x86.ActiveCfg = Debug|Win32
 		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Debug|x86.Build.0 = Debug|Win32
+		{E9D4AC92-8295-4FB4-BF7D-3FAF74B564E8}.Release|Any CPU.ActiveCfg = Release|Win32
```
Comment (ryanlai2) [inline Tools/WinMLRunner/WinMLRunner.sln]: With this change, When we specify Any CPU Debug, it'll build to Win32 Debug

File: Tools/WinMLRunner/WinMLRunner.sln (hunk 5 above)
Comment (ryanlai2) [reply Tools/WinMLRunner/WinMLRunner.sln]: Also, Debug|Any CPU.Build.0 is missing. If we click "Build solution" with the configuration: Any Debug | Any CPU, then WinMLRunner won't build.

File: Tools/WinMLRunner/BindingUtilities.h [hunk 8]
```diff
@@ -417,4 +417,96 @@ namespace BindingUtilities
...
+                    com_ptr<ITensorNative> itn = results.Lookup(desc.Name()).as<ITensorNative>();
+                    float* Tensor;
+                    uint32_t uCapacity;
+                    HRESULT(itn->GetBuffer(reinterpret_cast<BYTE**>(&Tensor), &uCapacity));
+                    hash = winrt::impl::hash_data(Tensor, uCapacity);
+
+                    UINT maxIndex = 0;
+                    auto maxValue = *Tensor;
+                    int size = uCapacity / sizeof(float_t);
+                    for (UINT i = 0; i < size; i++)
```
Comment (ryanlai2) [inline Tools/WinMLRunner/BindingUtilities.h]: Signed / unsigned Int mismatch comparison

**Comment by ryanlai2** [comment]
Can we get some test(s) added to test that this works properly?

File: Tools/WinMLRunner/BindingUtilities.h [hunk 9]
```diff
@@ -417,4 +417,96 @@ namespace BindingUtilities
...
+                TensorKind tensorKind = tensorDescriptor.TensorKind();
+
+                switch (tensorKind)
+                {
+                case TensorKind::String:
+                {
+                    com_ptr<ITensorNative> itn = results.Lookup(desc.Name()).as<ITensorNative>();
+                    std::string* Tensor;
+                    uint32_t uCapacity;
+                    HRESULT(itn->GetBuffer(reinterpret_cast<BYTE**>(&Tensor), &uCapacity));
```
Comment (smk2007) [inline Tools/WinMLRunner/BindingUtilities.h]: This is not supported for TensorString types. The GetBuffer method will return ERROR_INVALID_FUNCTION here.

File: Tools/WinMLRunner/BindingUtilities.h [hunk 10]
```diff
@@ -417,4 +417,96 @@ namespace BindingUtilities
...
+                }
+                break;
+                
+                case TensorKind::Float:
+                {
+                    com_ptr<ITensorNative> itn = results.Lookup(desc.Name()).as<ITensorNative>();
+                    float* Tensor;
+                    uint32_t uCapacity;
+                    HRESULT(itn->GetBuffer(reinterpret_cast<BYTE**>(&Tensor), &uCapacity));
+                    hash = winrt::impl::hash_data(Tensor, uCapacity);
```
Comment (smk2007) [inline Tools/WinMLRunner/BindingUtilities.h]: i dont think the winrt::impl namespace is safe here - the SDK team changes the impl internals quite frequently.

File: Tools/WinMLRunner/Main.cpp [hunk 11]
```diff
@@ -276,7 +280,7 @@ HRESULT EvaluateModel(
     // Run the binding + evaluate multiple times and average the results
     for (uint32_t i = 0; i < numIterations; i++)
     {
-        bool captureIterationPerf = (args.PerfCapture() && (!args.IgnoreFirstRun() || i > 0)) || (args.PerIterCapture());
+        bool captureIterationPerf = (args.PerfCapture() && (!args.IgnoreFirstRun() || i > 0)) || args.SaveTensor()  || args.PerIterCapture();;
```
Comment (ryanlai2) [inline Tools/WinMLRunner/Main.cpp]: Why do we need to capture performance when we want to save tensor output?

File: Tools/WinMLRunner/Main.cpp [hunk 12]
```diff
@@ -322,7 +332,7 @@ HRESULT EvaluateModels(
 
         try
         {
-            model = LoadModel(path, args.PerfCapture() || args.PerIterCapture(), output, args, 0);
+            model = LoadModel(path, args.PerfCapture() || args.SaveTensor() || args.PerIterCapture(), output, args, 0);
```
Comment (ryanlai2) [inline Tools/WinMLRunner/Main.cpp]: Why do we want to capture load model performance if we only want to save tensor?

File: Tools/WinMLRunner/Main.cpp [hunk 13]
```diff
@@ -348,7 +358,7 @@ HRESULT EvaluateModels(
                 {
                     for (auto deviceCreationLocation : deviceCreationLocations)
                     {
-                        if (args.PerfCapture() || args.PerIterCapture())
+                        if (args.PerfCapture() || args.SaveTensor() || args.PerIterCapture())
```
Comment (ryanlai2) [inline Tools/WinMLRunner/Main.cpp]: Why do we want to worry about performance when saving tensors?

**Comment by ryanlai2** [comment]
Hey @laks0209, here are some tensor test ideas that @pmbrown1055 and I spoke about:

Running WinMLRunner and outputting tensor CSV files and compare against an expected tensor CSV file to make sure values are correct.

With combinations of
- CPU / GPU
- Input Image as png, Input Image as CSV, Garbage data that is seeded so output tensor is deterministic

There may be a % threshold of error tolerance needed

**Comment by laks0209** [comment]
Thanks Ryan for the suggestions/corrections. I will update the code and add test cases as per the conversation.

**Comment by laks0209** [comment]
Hi @ryanlai2 , Please review the latest commit with the changes and the tests.

File: Tools/WinMLRunner/src/Run.cpp [hunk 14]
```diff
@@ -6,13 +6,118 @@
...
+	}
+
+	return model;
+}
 
 LearningModel LoadModel(const std::wstring path, bool capturePerf, OutputHelper& output, const CommandLineArgs& args, uint32_t iterationNum, Profiler<WINML_MODEL_TEST_PERF>& profiler)
 {
     LearningModel model = nullptr;
     output.PrintLoadingInfo(path);
+	model = LoadModelCryptography(path);
```
Comment (ryanlai2) [inline Tools/WinMLRunner/src/Run.cpp]: Why do we need this for saving output tensor results? Also, won't the below line of code:

``` model = LearningModel::LoadFromFilePath(path); ```

overwrite the loading of the model here?

File: Testing/WinMLRunnerTest/WinMLRunnerTest.cpp [hunk 15]
```diff
@@ -133,6 +180,8 @@ namespace WinMLRunnerTest
             Assert::AreEqual(static_cast<size_t>(2), GetOutputCSVLineCount());
         }
 
+
```
Comment (ryanlai2) [inline Testing/WinMLRunnerTest/WinMLRunnerTest.cpp]: Nit: Whitespace

File: Tools/WinMLRunner/src/BindingUtilities.h [hunk 16]
```diff
@@ -494,4 +515,89 @@ namespace BindingUtilities
...
+                            maxValue = val;
+                            maxIndex = i;
+                        }
+                    }
+                    iterationResult = "Index: " + std::to_string(maxIndex) + "; Value: " + std::to_string(maxValue);
+                    output.WriteTensorResultToCSV(Tensor, iterationNum, args, uCapacity, tempName);
+                }
+                break;
+
+                case TensorKind::Int64:
```
Comment (ryanlai2) [inline Tools/WinMLRunner/src/BindingUtilities.h]: Can we add a test to test this case?

File: Tools/WinMLRunner/src/Run.cpp [hunk 17]
```diff
@@ -6,118 +6,15 @@
...
-	return model;
-}
 
 LearningModel LoadModel(const std::wstring path, bool capturePerf, OutputHelper& output, const CommandLineArgs& args, uint32_t iterationNum, Profiler<WINML_MODEL_TEST_PERF>& profiler)
 {
     LearningModel model = nullptr;
     output.PrintLoadingInfo(path);
-	model = LoadModelCryptography(path);
-
+	model = LearningModel::LoadFromFilePath(path);
```
Comment (ryanlai2) [inline Tools/WinMLRunner/src/Run.cpp]: We already load the model below in line 25 so I think this line would be redundant. What do you think?

Comment (laks0209) [reply Tools/WinMLRunner/src/Run.cpp]: Yes thats right. Sorry. I have updated it.

File: Testing/WinMLRunnerTest/WinMLRunnerTest.cpp (hunk 15 above)
Comment (laks0209) [reply Testing/WinMLRunnerTest/WinMLRunnerTest.cpp]: Have made correction in the latest commit.

File: Tools/WinMLRunner/src/Run.cpp (hunk 14 above)
Comment (laks0209) [reply Tools/WinMLRunner/src/Run.cpp]: Have made correction in the latest commit.

File: Tools/WinMLRunner/src/OutputHelper.h [hunk 18]
```diff
@@ -305,21 +308,65 @@ class OutputHelper
...
     }
+
+    void SaveResult(uint32_t iterationNum, std::string result, int hashcode)
+    {
+        m_outputResult[iterationNum] = result;
+        m_outputTensorHash[iterationNum] = hashcode;
+    }
     
-    void SetDefaultCSVFileNamePerIteration()
+    void SetDefaultFolder()
```
Comment (ryanlai2) [inline Tools/WinMLRunner/src/OutputHelper.h]: Can we change this method name to capture the idea that it's for setting the folder for per iterations run data?

Maybe: "SetDefaultPerIterationFolder"

File: Tools/WinMLRunner/src/BindingUtilities.h (hunk 16 above)
Comment (ryanlai2) [reply Tools/WinMLRunner/src/BindingUtilities.h]: Is there a model that we can test for Int64? To verify that this code path works?

**Comment by laks0209** [comment]
Removed Int64 tensorkind. And have changed the naming of the folder. Please review the latest commit :-)

**Pull request merged**